src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,857 @@
+  - [Immutable types](#immutable-types)
+  - [Value types](#value-types)
+  - [Interaction with JsonPropertyNameAttribute](#interaction-with-jsonpropertynameattribute)
+- [Future](#future)
+- [Notes](#notes)
+
+
+# Terminology
+
+**Reference loops**: Also referred as circular references, loops occur when a property of a .NET object refers to the object itself, either directly (a -> a) or indirectly (a -> b -> a). They also occur when the element of an array refers to the array itself (arr[0] -> arr). Multiple occurrences of the same reference do not imply a cycle.
+
+**Preserve duplicated references**: Semantically represent objects and/or arrays that have been previously written, with a reference to them when found again in the object graph (using reference equality for comparison).
+
+**Metadata**: Extra properties on JSON objects and/or arrays (that may change their schema) to enable reference preservation when round-tripping. These additional properties are only meant to be understood by the `JsonSerializer`.  
+
+# Motivation
+
+Currently, there is no mechanism to avoid infinite loops while serializing .NET object instances that contain cycles nor to preserve references that round-trip when using `System.Text.Json`. The `JsonSerializer` throws a `JsonException` when a loop is found within the object graph.
+
+This is a heavily requested feature since it is considered by many as a very common scenario, especially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).
+
+The current solution to deal with cycles in the object graph while serializing is to rely on `MaxDepth` and throw a `JsonException` after it is exceeded. This was done to avoid perf overhead for cycle detection in the common case. The goal is to enable the new opt-in feature with minimal impact to existing performance.
+
+
+# Proposal
+
+```cs
+namespace System.Text.Json
+{
+    public partial class JsonSerializerOptions
+    {
+        public ReferenceHandling ReferenceHandling { get; set; } = ReferenceHandling.Default;
+    }
+}
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// This class defines the various ways the <see cref="JsonSerializer"/> 
+    /// can deal with references on Serialization and Deserialization.
+    /// </summary>
+    public sealed class ReferenceHandling
+    {
+        public static ReferenceHandling Default { get; }
+        public static ReferenceHandling Preserve { get; }
+        // TODO: decide if we keep or remove this option.
+        public static ReferenceHandling Ignore { get; }
+    }
+}
+```
+See also the [internal implementation details](https://gist.github.com/Jozkee/b0922ef609f7a942f00ac2c93a976ff1).
+
+## In depth
```

Comment:
nit: this title is a bit confusing - my mind first went to serialization depth (maybe it's just me) 

Consider "A closer look `Default`, `Preserve`, and `Ignore` reference handling options" or something...